OPHKIOS-184 ilmoittautumiset backend

saku-koodari opened the pull request:
## Yhteenveto

Asiakastiedot - sivun ilmoittautumiset - data

## Lisätiedot

_Tehdyt muutokset..._

## Huomautukset

- jos haluat testata tätä, niin huomioi, että tuon jälkeen kutsutaan metodeita, jotka heittää `NoImplemented` - virheen
https://github.com/Opetushallitus/kieli-ja-kaantajatutkinnot/blob/dev/backend/yki/src/main/java/fi/oph/yki/service/ClerkCustomerService.java#L38-L44



## Check-lista

- [ ] yarn.lock päivitetty
- [ ] Käännösten Excel-tiedosto päivitetty
- [ ] UI-muutoksia testattu eri selaimilla ja mobiilinäkymässä

saku-koodari committed:
YKI(Backend): wip - ilmoittautumiset

saku-koodari committed:
YKI(Backend): wip - Get data using JPA

saku-koodari committed:
YKI(Backend): Exam dates

saku-koodari committed:
YKI(Backend): Add test data

saku-koodari committed:
YKI(Backend): Exam location supports now different languages

saku-koodari committed:
YKI(Backend): Run lint

saku-koodari committed:
YKI(Backend): Add Exam Payment New to migrations

saku-koodari commented on backend/yki/src/main/java/fi/oph/yki/service/ClerkCustomerService.java:
```diff
@@ -35,8 +43,30 @@ private ClerkCustomerPersonDTO getClerkCustomerPersonDTO(String oid) throws Exce
-  private List<ClerkCustomerRegistrationDTO> getClerkCustomerRegistrationDTOs() {
-    throw new NotImplementedException("Get registrations from the database is not implemented yet.");
+  private List<ClerkCustomerRegistrationDTO> getClerkCustomerRegistrationDTOs(String oid) throws Exception {
+    return registrationRepository
```
Pullarin pihvi.

Asiakkaan ilmoittautumiset haetaan logiikalla:
- ilmoittautuminen haetaan `person_oid` - kentän perusteella
- haetaan ilmoittautumiset, joiden registration.kind != ‘ADMISSIION’

Siihen liitetyt tiedot:
- jos asiakas on maksanut ilmoittautumisen, se haetaan logiikalla: Uusin _maksettu_ maksu päivämäärä.
-  tutkintopäivä = `registration.exam_session.exam_date`
- ilmoittautumisen tila = `registration.state`
- ilmoittautumispvm = `registration.exam_session.exam_date.registration_start_date`, vai pitäisikö olla end date?

pkoivisto replied:
Ilmoittautumispvm ei ole saataville tuolta exam_sessionin puolelta, ne registration_start_date ja registration_end_date liittyy siihen, milloin ilmoittautumis**periodi** alkaa ja loppuu. Tällä hetkellä tieto ilmoittautumisesta asuu joko maksun (exam_payment_new) tai maksuttomuuden (free_registration) aikaleimoissa. Toki voisi myös tuonne registration-tauluun lisätä erillisen aikaleiman ilmoittautumisajalle.

saku-koodari replied:
Mä voisin tähän pullariin tehdä niin, että käytän olemassaolevaa aikaleimaa

jrkkp commented on backend/yki/src/main/java/fi/oph/yki/model/ExamSession.java:
```diff
@@ -37,4 +27,8 @@ public class ExamSession {
   private String level;
+
+  // TODO: Should i actually cascade all?
+  @OneToMany(mappedBy = "examSession", cascade = CascadeType.ALL, orphanRemoval = true)
```
Tuon cascade ja orphanRemoval voisi ottaa pois kun tietokanta huolehtii näistä

saku-koodari committed:
YKI(Backend): Remove cascades and orphan definitions

jrkkp commented on frontend/packages/yki/src/tests/msw/fixtures/customerDetails.ts:
```diff
@@ -5,17 +5,68 @@ import {
+  {
+    schoolName: 'Testipaikan nimi',
+    municipality: 'Kajaani',
+    lang: 'fi',
```
Tässä varmaan pitäisi olla schoolName sijasta name. Bäkkärissä ei myöskään ole tuota lang tietoa (`ClerkExamLocationDTO`), niin se varmaan pitänee lisätä sinne?

saku-koodari replied:
Hyvä pointti, tuo oli jostain syystä lipsahtanut silmieni alta

saku-koodari committed:
YKI(Backend): Filter out registrations that are not queued

saku-koodari committed:
YKI(Backend): Fixes
 - RegistrationDate is either the latest ExamPayment.PaidAt or FreeRegistration.CreatedAt
 - Merge registrations, queuedExams and pastExams

saku-koodari committed:
YKI(Backend): Add Language to ExamSessionLocation

jrkkp commented on frontend/packages/yki/src/utils/serialization.ts:
```diff
@@ -437,40 +439,96 @@ export class SerializationUtils {
+
+  static mapExamLocation(
+    examLocation: {
+      schoolName: string;
```
`schoolName` -> `name`

saku-koodari replied:
Otin nämä muutokset pois, niin teen sen siinä fronttipullarissa

saku-koodari commented on frontend/packages/yki/src/utils/serialization.ts:
```diff
@@ -437,40 +439,96 @@ export class SerializationUtils {
+              }),
+        },
+        queueSpotOffered: {
+          // TODO: How to determine queue spot offered?
```
Mistä saadaan `Jonopaikkaa tarjottu`?

saku-koodari commented on frontend/packages/yki/src/utils/serialization.ts:
```diff
@@ -437,40 +439,96 @@ export class SerializationUtils {
+        exam: registration.exam,
+        examinationDate: dayjs(registration.examinationDate),
+        examLocation: this.mapExamLocation(registration.examLocation),
+        // TODO: How to determine past exam state? (REVIEWED, CANCELLED, REGISTERED)
```
Mistä saadaan tälle arvo?

jrkkp commented on backend/yki/src/main/java/fi/oph/yki/service/ClerkCustomerService.java:
```diff
@@ -35,24 +46,41 @@ private ClerkCustomerPersonDTO getClerkCustomerPersonDTO(String oid) throws Exce
+    return registrationRepository
+      .getByPersonOid(oid)
+      .stream()
+      .filter(r -> r.getKind() != RegistrationKind.QUEUE)
```
Tätä filtteröintiä ei varmaan tarvita

pkoivisto commented on backend/yki/src/main/java/fi/oph/yki/model/ExamPayment.java:
```diff
@@ -0,0 +1,31 @@
+  private PaymentState state;
+
+  @Column(name = "paid_at")
+  private LocalDate paidAt;
```
`paid_at` on kannassa tyyppiä `timestamp with time zone`, eli oisko tässä `LocalDateTime` itse asiassa tarkempi?

saku-koodari committed:
YKI(Backend): Change paidAt type to LocalDateTime

saku-koodari committed:
YKI(Backend): Remove filter by kind
